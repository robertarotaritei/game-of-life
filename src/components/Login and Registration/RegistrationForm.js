import React from 'react';
import InputField from './InputField';
import SubmitButton from './SubmitButton';
import { Redirect } from 'react-router-dom';
import UserStore from '../../stores/UserStore';
import { runInAction } from 'mobx';

class RegistrationForm extends React.Component {

  constructor() {
    super();
    this.state = {
      username: '',
      password: '',
      passwordConfirmation: '',
      redirectDashboard: false
    }

    runInAction(() => {
      UserStore.key = sessionStorage.getItem('key');
      UserStore.loading = false;
    });
  }

  setInputValue(property, val) {
    val = val.trim();
    if (val.length > 15) {
      return;
    }
    this.setState({
      [property]: val
    })
  }

  setRedirect = () => {
    this.setState({
      redirect: true
    })
  }

  setRedirectDashboard = () => {
    this.setState({
      redirectDashboard: true
    })
  }

  renderRedirect = () => {
    if (this.state.redirect) {
      return <Redirect to='/login' />
    }
    if (this.state.redirectDashboard) {
      return <Redirect to='/' />
    }
  }

  async doRegister() {
    if (!this.state.username) {
      return;
    }

    if (!this.state.password) {
      return;
    }

    if (!this.state.passwordConfirmation) {
      return;
    }

    runInAction(() => {
      UserStore.loading = true;
    });

    let error = '';
    error = this.state.password !== this.state.passwordConfirmation ? 'Passwords do not match' : '';
    let res = await fetch(`/credentials/user/verifyusername?username=${this.state.username}`)
    let result = await res.json();

    runInAction(() => {
      UserStore.loading = false;
    });

    error = result.title !== "Not Found" ? 'Username already exists' : error;
    if (error === '') {
      runInAction(() => {
        UserStore.loading = true;
      });

      var newUser = {
        username: this.state.username,
        password: this.state.password
      };
      fetch('/credentials/user', {
        method: 'POST',
        mode: 'cors',
        headers: {
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(newUser)
      })

      alert("New user created");
      this.setRedirect();

      runInAction(() => {
        UserStore.loading = false;
      });
    }
    else {
      alert(error);
    }
  }

  render() {
    return (
      <div className="container">
        <div className='registrationFormPolly'>
          <div className="loginForm">
            {this.renderRedirect()}
            <div style={{ textAlign: 'center', color: '#17c5fa', marginTop: '25px' }}>
              <p>
                Register
          </p>
            </div>
            <InputField
              className='username'
              type='text'
              placeholder='Username'
              value={this.state.username ? this.state.username : ''}
              onChange={(val) => this.setInputValue('username', val)}
            />
            <InputField
              className='password'
              type='password'
              placeholder='Password'
              value={this.state.password ? this.state.password : ''}
              onChange={(val) => this.setInputValue('password', val)}
            />
            <InputField
              className='repeatPassword'
              type='password'
              placeholder='Repeat Password'
              value={this.state.passwordConfirmation ? this.state.passwordConfirmation : ''}
              onChange={(val) => this.setInputValue('passwordConfirmation', val)}
            />
            <div style={{ textAlign: 'center' }}>
              <SubmitButton
                text='REGISTER'
                onClick={() => this.doRegister()}
              />
            </div>
            <div style={{ display: 'flex', marginBottom: '30px', marginTop: '30px' }}>
              <p className="textAnimationBig">
                Already have an account?
              </p>
              <button className='button' onClick={this.setRedirect} >
                Log in
              </button>
            </div>
            <div style={{ display: 'flex', marginBottom: '10px' }}>
              <p className="textAnimationSmall" style={{ fontSize: '1.2em' }}>
                Don't want to log in?
              </p>
              <button className='button' onClick={this.setRedirectDashboard} style={{ fontSize: '1.2em' }}>
                Go back to the game
              </button>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

export default RegistrationForm;