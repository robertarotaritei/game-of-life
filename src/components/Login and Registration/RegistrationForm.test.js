import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import RegistrationForm from './RegistrationForm';
import { shallow } from "enzyme";

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve({ title: 'Not Found' }),
  })
);

beforeEach(() => {
  fetch.mockClear();
});

it('RegistrationForm clicks', () => {
  jest.spyOn(window, 'alert').mockImplementation(() => { });
  const wrapper = shallow(<RegistrationForm />);
  wrapper.find('SubmitButton').simulate('click');
  wrapper.find('.username').simulate('change', 'usernametoolong1234');
  wrapper.find('SubmitButton').simulate('click');
  wrapper.find('.username').simulate('change', 'username');
  wrapper.find('SubmitButton').simulate('click');
  wrapper.find('.password').simulate('change', 'password');
  wrapper.find('SubmitButton').simulate('click');
  wrapper.find('.repeatPassword').simulate('change', 'password1');
  wrapper.find('SubmitButton').simulate('click');
  wrapper.find('.repeatPassword').simulate('change', 'password');
  wrapper.find('SubmitButton').simulate('click');
});

test("RegistrationForm renders without crashing", () => {
  shallow(<RegistrationForm />);
});

test('Render RegistrationForm Text', () => {
  const { getByText } = render(<RegistrationForm />);
  const linkElement = getByText(/Already have an account/i);
  expect(linkElement).toBeInTheDocument();
});

it('RegistrationForm renders correctly', () => {
  const tree = renderer
    .create(<RegistrationForm />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});