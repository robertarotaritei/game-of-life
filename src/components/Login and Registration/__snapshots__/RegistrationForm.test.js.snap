// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="registrationFormPolly"
  >
    <div
      className="loginForm"
    >
      <div
        style={
          Object {
            "color": "#17c5fa",
            "marginTop": "25px",
            "textAlign": "center",
          }
        }
      >
        <p>
          Register
        </p>
      </div>
      <div
        className="inputPolly"
        style={
          Object {
            "marginTop": "30px",
          }
        }
      >
        <input
          onChange={[Function]}
          placeholder="Username"
          type="text"
          value=""
        />
      </div>
      <div
        className="inputPolly"
        style={
          Object {
            "marginTop": "30px",
          }
        }
      >
        <input
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
      <div
        className="inputPolly"
        style={
          Object {
            "marginTop": "30px",
          }
        }
      >
        <input
          onChange={[Function]}
          placeholder="Repeat Password"
          type="password"
          value=""
        />
      </div>
      <div
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <div
          className="submitButton"
          style={
            Object {
              "marginTop": "10px",
            }
          }
        >
          <button
            className="btn"
            onClick={[Function]}
          >
            <span
              className="btn__content"
            >
              REGISTER
            </span>
            <span
              className="btn__glitch"
            />
          </button>
        </div>
      </div>
      <div
        style={
          Object {
            "display": "flex",
            "marginBottom": "30px",
            "marginTop": "30px",
          }
        }
      >
        <p
          className="textAnimationBig"
        >
          Already have an account?
        </p>
        <button
          className="button"
          onClick={[Function]}
        >
          Log in
        </button>
      </div>
      <div
        style={
          Object {
            "display": "flex",
            "marginBottom": "10px",
          }
        }
      >
        <p
          className="textAnimationSmall"
          style={
            Object {
              "fontSize": "1.2em",
            }
          }
        >
          Don't want to log in?
        </p>
        <button
          className="button"
          onClick={[Function]}
          style={
            Object {
              "fontSize": "1.2em",
            }
          }
        >
          Go back to the game
        </button>
      </div>
    </div>
  </div>
</div>
`;
